dali/operators/generic/one_hot.cuh

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstdint>
+
+namespace dali {
+
+namespace detail {
+
+struct SampleDesc {
+  uint64_t inner_vol, output_vol, inner_vol_classes;
+  void *out = nullptr;
+  const void *in = nullptr;
+};
+
+template <typename OutputType, typename InputType>
+__global__ void PopulateOneHot(OutputType on_value, OutputType off_value,
+                               const SampleDesc *samples) {
+  uint64_t out_index = static_cast<uint64_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const auto &sample = samples[blockIdx.y];
+  if (out_index >= sample.output_vol) {
+    return;
+  }
+  auto *out = static_cast<OutputType*>(sample.out);
+  auto *in = static_cast<const InputType*>(sample.in);
+  uint64_t i = out_index / sample.inner_vol_classes;
```

Comment:
consider using ```fast_div<uint64_t>``` (example in slice_gpu.cuh)

Reply:
I think it was tested and it was slower.

Reply:
Yeah, it performed slight slower, this is what I got using fast_div for simultaneously getting modulo and div for division by inner_vol.

                                                                                                                   fast_div                             regular
OneHotTestParams<int, int, 1, 256, 0, 1024, 1024>,:                   Throughput: 292.507 GB/s             292.198 GB/s
OneHotTestParams<int, int, 1, 256, 1, 1024, 1024>,:                   Throughput: 411.535 GB/s             442.841 GB/s
OneHotTestParams<int, int, 1, 256, 2, 1024, 1024>,:                   Throughput: 408.046 GB/s             428.516 GB/s
OneHotTestParams<int, int64_t, 1, 256, 0, 1024, 1024>,:            Throughput: 204.88 GB/s              204.66 GB/s
OneHotTestParams<int, int64_t, 1, 256, 1, 1024, 1024>,:            Throughput: 413.583 GB/s             443.527 GB/s
OneHotTestParams<int, int64_t, 1, 256, 2, 1024, 1024>,:            Throughput: 410.612 GB/s             429.418 GB/s
OneHotTestParams<int64_t, int64_t, 1, 256, 0, 1024, 1024>,:     Throughput: 298.738 GB/s             298.621 GB/s
OneHotTestParams<int64_t, int64_t, 1, 256, 1, 1024, 1024>,:     Throughput: 634.285 GB/s             633.726 GB/s
OneHotTestParams<int64_t, int64_t, 1, 256, 2, 1024, 1024>,:     Throughput: 629.212 GB/s             629.856 GB/s
OneHotTestParams<int8_t, int8_t, 1, 256, 0, 1024, 1024>,:         Throughput: 100.574 GB/s             112.423 GB/s
OneHotTestParams<int8_t, int8_t, 1, 256, 1, 1024, 1024>,:         Throughput: 100.88 GB/s              112.152 GB/s
OneHotTestParams<int8_t, int8_t, 1, 256, 2, 1024, 1024>,:         Throughput: 100.576 GB/s             111.05 GB/s
OneHotTestParams<int, int, 4, 8, 0, 32, 32, 32, 32, 32>,:             Throughput: 326.758 GB/s             326.245 GB/s
OneHotTestParams<int, int, 4, 8, 1, 32, 32, 32, 32, 32>,:             Throughput: 326.828 GB/s             326.34 GB/s
OneHotTestParams<int, int, 4, 8, 2, 32, 32, 32, 32, 32>,:             Throughput: 457.538 GB/s             485.431 GB/s
OneHotTestParams<int, int, 4, 8, 3, 32, 32, 32, 32, 32>,:             Throughput: 447.502 GB/s             456.19 GB/s
OneHotTestParams<int, int, 4, 8, 4, 32, 32, 32, 32, 32>,:             Throughput: 449.084 GB/s             460.766 GB/s
OneHotTestParams<int, int, 4, 8, 5, 32, 32, 32, 32, 32>,:             Throughput: 447.849 GB/s             460.415 GB/s
OneHotTestParams<int, int, 4, 1024 * 256, 0, 1024>,:                 Throughput: 414.606 GB/s             459.603 GB/s
OneHotTestParams<int, int, 4, 1024 * 256, 1, 1024>,:                 Throughput: 414.489 GB/s             457.48 GB/s
OneHotTestParams<int, int, 16, 64, 0, 1024, 128>,:                     Throughput: 414.887 GB/s             459.284 GB/s
OneHotTestParams<int, int, 16, 64, 1, 1024, 128>,:                    Throughput: 406.263 GB/s             422.71 GB/s
OneHotTestParams<int, int, 16, 64, 2, 1024, 128>:                     Throughput: 410.158 GB/s             430.671 GB/s

Reply:
@mzient Any idea why that might be the case?